File: kitty-api/kitty-material-api/src/main/java/icu/jiapeng/kitty/material/permission/constants/MaterialPermissionCode.java
package icu.jiapeng.kitty.material.permission.constants;

/**
 * material 模块权限编码。
 */
public interface MaterialPermissionCode {

    /**
     * 栏目树查看。
     */
    String MATERIAL_CATALOG_TREE_VIEW = "material:catalog:tree:view";

    /**
     * 资源列表查看。
     */
    String MATERIAL_RESOURCE_LIST_VIEW = "material:resource:list:view";

    /**
     * 资源创建。
     */
    String MATERIAL_RESOURCE_CREATE = "material:resource:create";

    /**
     * 资源更新。
     */
    String MATERIAL_RESOURCE_UPDATE = "material:resource:update";

    /**
     * 编目模板管理（创建/绑定/启停）。
     */
    String MATERIAL_METADATA_TEMPLATE_MANAGE = "material:metadata:template:manage";

    /**
     * 编目字段定义管理。
     */
    String MATERIAL_METADATA_FIELD_MANAGE = "material:metadata:field:manage";

    /**
     * 转码策略与栏目绑定管理。
     */
    String MATERIAL_TRANSCODE_POLICY_MANAGE = "material:transcode:policy:manage";

    /**
     * 站内信发送。
     */
    String MATERIAL_MESSAGE_SEND = "material:message:send";

    /**
     * 站内信查看（收件箱、SSE）。
     */
    String MATERIAL_MESSAGE_READ = "material:message:read";

    /**
     * 通用审核提交。
     */
    String MATERIAL_REVIEW_SUBMIT = "material:review:submit";

    /**
     * 通用审核处理（通过/拒绝）。
     */
    String MATERIAL_REVIEW_APPROVE = "material:review:approve";

    /**
     * 栏目权限管理。
     */
    String MATERIAL_CATALOG_PERMISSION_EDIT = "material:catalog:permission:edit";

    /**
     * 栏目权限的查看
     */
    String MATERIAL_CATALOG_PERMISSION_VIEW = "material:catalog:permission:view";
    /**
     * 新建栏目权限
     */
    String MATERIAL_CATALOG_CREATE = "material:catalog:create";

    /**
     * 栏目信息修改（重命名、移动等）。
     */
    String MATERIAL_CATALOG_UPDATE = "material:catalog:update";

    /**
     * 栏目删除。
     */
    String MATERIAL_CATALOG_DELETE = "material:catalog:delete";

    /**
     * 文件存储配置管理（增删改查 kt_file_storage，含 S3 等引擎）。
     */
    String MATERIAL_STORAGE_MANAGE = "material:storage:manage";
}

File: frontend/packages/mam/src/constants/material_permission_code.ts
/**
 * 与后端 `icu.jiapeng.kitty.material.permission.constants.MaterialPermissionCode` 字符串一致。
 */
export const MaterialPermissionCode = {
  MATERIAL_CATALOG_TREE_VIEW: 'material:catalog:tree:view',
  MATERIAL_RESOURCE_LIST_VIEW: 'material:resource:list:view',
  MATERIAL_RESOURCE_CREATE: 'material:resource:create',
  MATERIAL_RESOURCE_UPDATE: 'material:resource:update',
  MATERIAL_METADATA_TEMPLATE_MANAGE: 'material:metadata:template:manage',
  MATERIAL_METADATA_FIELD_MANAGE: 'material:metadata:field:manage',
  MATERIAL_TRANSCODE_POLICY_MANAGE: 'material:transcode:policy:manage',
  MATERIAL_MESSAGE_SEND: 'material:message:send',
  MATERIAL_MESSAGE_READ: 'material:message:read',
  MATERIAL_REVIEW_SUBMIT: 'material:review:submit',
  MATERIAL_REVIEW_APPROVE: 'material:review:approve',
  MATERIAL_CATALOG_PERMISSION_EDIT: 'material:catalog:permission:edit',
  MATERIAL_CATALOG_PERMISSION_VIEW: 'material:catalog:permission:view',
  MATERIAL_CATALOG_CREATE: 'material:catalog:create',
  MATERIAL_CATALOG_UPDATE: 'material:catalog:update',
  MATERIAL_CATALOG_DELETE: 'material:catalog:delete',
  MATERIAL_STORAGE_MANAGE: 'material:storage:manage'
} as const

export type MaterialPermissionCodeKey = keyof typeof MaterialPermissionCode
export type MaterialPermissionCodeValue =
  (typeof MaterialPermissionCode)[MaterialPermissionCodeKey]

/**
 * 与后端 `CatalogPermission` 枚举顺序、含义一致；用于 `kt_catalog_permission` / `requireOnCatalog` 维度。
 */
export const CATALOG_DIMENSION_PERMISSION_CODES: readonly MaterialPermissionCodeValue[] = [
  MaterialPermissionCode.MATERIAL_CATALOG_TREE_VIEW,
  MaterialPermissionCode.MATERIAL_CATALOG_UPDATE,
  MaterialPermissionCode.MATERIAL_CATALOG_DELETE,
  MaterialPermissionCode.MATERIAL_CATALOG_CREATE,
  MaterialPermissionCode.MATERIAL_METADATA_TEMPLATE_MANAGE,
  MaterialPermissionCode.MATERIAL_RESOURCE_LIST_VIEW,
  MaterialPermissionCode.MATERIAL_RESOURCE_CREATE,
  MaterialPermissionCode.MATERIAL_RESOURCE_UPDATE,
  MaterialPermissionCode.MATERIAL_TRANSCODE_POLICY_MANAGE,
  MaterialPermissionCode.MATERIAL_REVIEW_SUBMIT,
  MaterialPermissionCode.MATERIAL_REVIEW_APPROVE
]

/**
 * 迁移前 kt_catalog_permission 使用的短码，与 canonical 等价（与后端兼容逻辑一致）。
 */
export const LEGACY_CATALOG_PERMISSION_BY_CANONICAL: Readonly<
  Partial<Record<MaterialPermissionCodeValue, readonly string[]>>
> = {
  [MaterialPermissionCode.MATERIAL_CATALOG_TREE_VIEW]: ['view'],
  [MaterialPermissionCode.MATERIAL_CATALOG_UPDATE]: ['edit'],
  [MaterialPermissionCode.MATERIAL_CATALOG_CREATE]: ['add'],
  [MaterialPermissionCode.MATERIAL_CATALOG_DELETE]: ['delete']
}

/** 判断节点 permissionCodes 是否具备某栏目维度权限（含历史短码） */
export function catalogNodeHasPermission(
  permissionCodes: string[] | undefined,
  canonical: MaterialPermissionCodeValue
): boolean {
  if (!permissionCodes?.length) {
    return false
  }
  if (permissionCodes.includes(canonical)) {
    return true
  }
  const legacy = LEGACY_CATALOG_PERMISSION_BY_CANONICAL[canonical]
  if (!legacy?.length) {
    return false
  }
  return legacy.some((c) => permissionCodes.includes(c))
}
